# Data Cleaning with Pandas

In [39]:
import pandas as pd

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it: `brew install wget`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the tile that you want to unzip.


You can use tab completion (press the `tab` key after the first three letters) to fill in the names, including spaces. This will 

In [40]:
sales_df = pd.read_csv('data/EXTR_RPSale.csv')
sales_df.head()

ExciseTaxNbr   Major Minor DocumentDate  SalePrice    RecordingNbr Volume  \
0       2687551  138860   110   08/21/2014     245000  20140828001436          
1       1235111  664885    40   07/09/1991          0  199203161090      071   
2       2704079  423943    50   10/11/2014          0  20141205000558          
3       2584094  403700   715   01/04/2013          0  20130110000910          
4       1056831  951120   900   04/20/1989      85000  198904260448      117   

  Page PlatNbr PlatType     ...     PropertyType PrincipalUse SaleInstrument  \
0                           ...                3            6              3   
1  001  664885        C     ...                3            0             26   
2                           ...                3            6             15   
3                           ...                3            6             15   
4  053  951120        P     ...                3            0              2   

  AFForestLand  AFCurrentUseLand  AFNonProfitUse  AFHistoricProperty  \
0            N                 N               N                   N   
1            N                 N               N                   N   
2            N                 N               N                   N   
3            N                 N               N                   N   
4            N                 N               N                   N   

  SaleReason PropertyClass SaleWarning  
0          1             8              
1         18             3          11  
2         18             8    18 31 51  
3         11             8    18 31 38  
4          1             9          49  

[5 rows x 24 columns]

### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [41]:
sales_df.head().T

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [42]:
#Subset the dataframe by the columns we are interested in 
sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']]
sales_df.head()

Major Minor DocumentDate  SalePrice
0  138860   110   08/21/2014     245000
1  664885    40   07/09/1991          0
2  423943    50   10/11/2014          0
3  403700   715   01/04/2013          0
4  951120   900   04/20/1989      85000

In [43]:
sales_df.info()
#The objects are something we are going to have to deal with later

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.5+ MB


In [44]:
bldg_df = pd.read_csv('data/Residential Building.zip')

### Another warning! Which column has index 11?

In [45]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [46]:
bldg_df.head().T

0                            1  \
Major                                    4300                         4610   
Minor                                     167                          399   
BldgNbr                                     1                            1   
NbrLivingUnits                              1                            1   
Address             15223   40TH AVE S  98188  4431   FERNCROFT RD   98040   
BuildingNumber                          15223                        4431    
Fraction                                                                     
DirectionPrefix                                                              
StreetName          40TH                         FERNCROFT                   
StreetType                             AVE                          RD       
DirectionSuffix                            S                                 
ZipCode                                 98188                        98040   
Stories                                     1                            2   
BldgGrade                                   6                            9   
BldgGradeVar                                0                            0   
SqFt1stFloor                              990                          980   
SqFtHalfFloor                               0                            0   
SqFt2ndFloor                                0                         1140   
SqFtUpperFloor                              0                            0   
SqFtUnfinFull                               0                            0   
SqFtUnfinHalf                               0                            0   
SqFtTotLiving                             990                         2950   
SqFtTotBasement                           990                          940   
SqFtFinBasement                             0                          830   
FinBasementGrade                            0                            9   
SqFtGarageBasement                          0                            0   
SqFtGarageAttached                          0                          480   
DaylightBasement                                                         Y   
SqFtOpenPorch                              80                            0   
SqFtEnclosedPorch                           0                            0   
SqFtDeck                                    0                          600   
HeatSystem                                  6                            5   
HeatSource                                  1                            2   
BrickStone                                  0                            0   
ViewUtilization                                                          N   
Bedrooms                                    2                            3   
BathHalfCount                               0                            1   
Bath3qtrCount                               0                            0   
BathFullCount                               1                            2   
FpSingleStory                               0                            0   
FpMultiStory                                0                            0   
FpFreestanding                              0                            2   
FpAdditional                                0                            0   
YrBuilt                                  1922                         1978   
YrRenovated                                 0                            0   
PcntComplete                                0                            0   
Obsolescence                                0                            0   
PcntNetCondition                            0                            0   
Condition                                   3                            4   
AddnlCost                                   0                            0   

                                              2                  

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [47]:
#subset to the features I'm intersted in 
bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']]

In [48]:
bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511412 entries, 0 to 511411
Data columns (total 4 columns):
Major            511412 non-null int64
Minor            511412 non-null int64
SqFtTotLiving    511412 non-null int64
ZipCode          468372 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [49]:
#I want to merge major and minor columns because that is a unique identifer for a property 
#However, I will see an error because I cannot combine two types of data - one is an object and the other is numeric
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [50]:
#try to convert the stack trace - like saying error - to numeric so we have all the same data types 
# better way to say it: converting the major column within sales_df from type object to type float 64
sales_df['Major'] = pd.to_numeric(sales_df['Major'])

ValueError: Unable to parse string "      " at position 935519

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [ ]:
# The single question mark means "show me the docstring"
# Some of the values are essentially NA, an empty string -- what should I do
# I have to make a decision of what to turn the empty string into
# question mark gives me my choices, or docstring  
# coerce parces into floats and not int. We want to use a float not int
pd.to_numeric?

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [51]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'], errors='coerce')

Did it work?

In [52]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: float64(1), int64(1), object(2)
memory usage: 61.5+ MB


It worked! Let's do the same thing with the `Minor` parcel number.

In [53]:
sales_df['Minor'] = pd.to_numeric(sales_df['Minor'], errors='coerce')

In [54]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 61.5+ MB


Now, let's try our join again.

In [55]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

In [56]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [57]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1437522 entries, 0 to 1437521
Data columns (total 6 columns):
Major            1437522 non-null float64
Minor            1437522 non-null float64
DocumentDate     1437522 non-null object
SalePrice        1437522 non-null int64
SqFtTotLiving    1437522 non-null int64
ZipCode          1322148 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 76.8+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [58]:
sales_data.loc[sales_data['ZipCode'].isna()].head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
99   858140.0  376.0   05/22/2012          0            900     NaN
100  858140.0  376.0   11/28/2017          0            900     NaN
152  720319.0  520.0   11/20/2013     699950           2840     NaN
153  720319.0  520.0   09/16/2013          0           2840     NaN
163  894677.0  240.0   12/21/2016     818161           2450     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [128]:
#Detects the missing values in ZipCode with isna() and drops them  
sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :]
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0  138860.0  110.0   08/21/2014     245000           1490   98002   
1  138860.0  110.0   06/12/1989     109300           1490   98002   
2  138860.0  110.0   01/16/2005      14684           1490   98002   
6  423943.0   50.0   07/15/1999      96000            960   98092   
7  423943.0   50.0   01/08/2001     127500            960   98092   

   PricePerSqFt  
0    164.429530  
1     73.355705  
2      9.855034  
6    100.000000  
7    132.812500

# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [60]:
#Detects the missing values in SalePrice with isna() and drops them 
sales_data = sales_data.loc[~sales_data['SalePrice'].isna(), :]
sales_data.head()

#Detects the zeroes in SalePrice and drops them 
sales_data = sales_data.loc[sales_data['SalePrice']!=0]
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
6  423943.0   50.0   07/15/1999      96000            960   98092
7  423943.0   50.0   01/08/2001     127500            960   98092

In [61]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 867416 entries, 0 to 1437520
Data columns (total 6 columns):
Major            867416 non-null float64
Minor            867416 non-null float64
DocumentDate     867416 non-null object
SalePrice        867416 non-null int64
SqFtTotLiving    867416 non-null int64
ZipCode          867416 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 46.3+ MB


In [62]:
#Detects the missing values in SqFtTotLiving with isna() and drops them 
sales_data = sales_data.loc[~sales_data['SqFtTotLiving'].isna(), :]
sales_data.head()

#Detects the zeroes in SalePrice and drops them 
sales_data = sales_data.loc[sales_data['SqFtTotLiving']!=0]
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
6  423943.0   50.0   07/15/1999      96000            960   98092
7  423943.0   50.0   01/08/2001     127500            960   98092

In [63]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 867407 entries, 0 to 1437520
Data columns (total 6 columns):
Major            867407 non-null float64
Minor            867407 non-null float64
DocumentDate     867407 non-null object
SalePrice        867407 non-null int64
SqFtTotLiving    867407 non-null int64
ZipCode          867407 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 46.3+ MB


### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [64]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
# def is_integer(x):
#     try:
#         _ = int(x)
#     except ValueError:
#         return False
#     return True

def short_zip(x): 
    if type(x) is str:
        if len(x) > 5: 
            return x[0:5]
        else: 
            return x 

sales_data['ZipCode'] = sales_data['ZipCode'].apply(short_zip)
sales_data['ZipCode'].head()

0    98002
1    98002
2    98002
6    98092
7    98092
Name: ZipCode, dtype: object

### 3. Add a column for PricePerSqFt



In [105]:
sales_data["PricePerSqFt"] = sales_data["SalePrice"] / sales_data["SqFtTotLiving"]

sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
0  138860.0  110.0   08/21/2014     245000           1490   98002   
1  138860.0  110.0   06/12/1989     109300           1490   98002   
2  138860.0  110.0   01/16/2005      14684           1490   98002   
6  423943.0   50.0   07/15/1999      96000            960   98092   
7  423943.0   50.0   01/08/2001     127500            960   98092   

   PricePerSqFt  
0    164.429530  
1     73.355705  
2      9.855034  
6    100.000000  
7    132.812500

### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

In [129]:
#Testing to get one year value from the DocumentDate series 
dates = sales_data.loc[:,"DocumentDate"]
type(dates[0][-4:])

#create a function to find all the dates that end in the year 2019  
def newest_data(date):
    return date.endswith('2019')
    
#apply the function to the location, column Document Date in the sales data dataframe
sales_data_2019 = sales_data.loc[sales_data['DocumentDate'].apply(newest_data), : ]

#see a sample of results 
sales_data_2019.head()       

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
605   214090.0  110.0   03/06/2019     390000           1780   98042   
651   367890.0   30.0   03/13/2019     923000           2550   98144   
868   350160.0  125.0   02/06/2019     935000           2460   98117   
1063  882790.0  870.0   02/15/2019     890000           2380   98105   
1341   30200.0  645.0   03/04/2019     380000           1750   98001   

      PricePerSqFt  
605     219.101124  
651     361.960784  
868     380.081301  
1063    373.949580  
1341    217.142857

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

In [134]:
#I am not sure why the following does not work 
# def seattle_city_only(zipcode):
#     seattle_zips = ['98101', '98102', '98103', '98104', '98105', '98106', '98107', '98108', '98109', '98112', '98115', '98116', '98117', '98118', '98119', '98121', '98122', '98125', '98126', '98133', '98134', '98136', '98144', '98146', '98154', '98164', '98174', '98177', '98178', '98195', '98199']
#     if zipcode in seattle_zips:
#         return zipcode

# city_of_seattle_2019_data = sales_data_2019.loc[sales_data_2019['ZipCode'].apply(seattle_city_only),:]
# city_of_seattle_2019_data.head()

#######
seattle_zips = ['98101', '98102', '98103', '98104', '98105', '98106', '98107', '98108', '98109', '98112', '98115', '98116', '98117', '98118', '98119', '98121', '98122', '98125', '98126', '98133', '98134', '98136', '98144', '98146', '98154', '98164', '98174', '98177', '98178', '98195', '98199']
city_of_seattle_2019_data = sales_data_2019.loc[sales_data_2019.ZipCode.isin(seattle_zips)]
city_of_seattle_2019_data.head()


Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
651   367890.0   30.0   03/13/2019     923000           2550   98144   
868   350160.0  125.0   02/06/2019     935000           2460   98117   
1063  882790.0  870.0   02/15/2019     890000           2380   98105   
2821  945920.0  125.0   01/29/2019    1750000           1400   98118   
5359  186240.0  105.0   03/19/2019     380000            800   98117   

      PricePerSqFt  
651     361.960784  
868     380.081301  
1063    373.949580  
2821   1250.000000  
5359    475.000000

### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!

In [149]:
print("The mean price per square foot for a house sold in Seattle in 2019 is $"+str(round(city_of_seattle_2019_data.PricePerSqFt.mean(),2))+" per square foot.")


The mean price per square foot for a house sold in Seattle in 2019 is $505.96 per square foot.
